// bundle/bundleData.js
// age: [0-17, 18-64, 65+]
// isStudent: [true, false]
// income: [0, 1-12000, 12001-40000, 40000+]

// 'currentAccount', 'currentAccountPlus', 'juniorSaverAccount', 'studentAccount', 'debitCard', 'creditCard', 'goldCreditCard'

module.exports = {
    juniorSaver: {
        name: 'Junior Saver',
        account: 'juniorSaverAccount',
        card: [],
        rules: {
            age: [0],
            isStudent: [true, false],
            income: [0, 1, 2, 3]
        },
        value: 0
    },
    student: {
        name: 'Student',
        account: 'studentAccount',
        card: [
            'debitCard',
            'creditCard'
        ],
        rules: {
            age: [1, 2, 3],
            isStudent: [true],
            income: [0, 1, 2, 3]
        },
        value: 0
    },
    classic: {
        name: 'Classic',
        account: 'currentAccount',
        card: [
            'debitCard'
        ],
        rules: {
            age: [1, 2, 3],
            isStudent: [true, false],
            income: [1, 2, 3]
        },
        value: 1
    },
    classicPlus: {
        name: 'Classic Plus',
        account: 'currentAccount',
        card: [
            'debitCard',
            'creditCard'
        ],
        rules: {
            age: [1, 2, 3],
            isStudent: [true, false],
            income: [2, 3]
        },
        value: 2
    },
    gold: {
        name: 'Gold',
        account: 'currentAccountPlus',
        card: [
            'debitCard',
            'goldCreditCard'
        ],
        rules: {
            age: [1, 2, 3],
            isStudent: [true, false],
            income: [3]
        },
        value: 3
    }
};
